### PR TITLE
storage: fix block rows not match when filter column is the first non-empty column in the block

### LINKED ISSUE #9472: query count(*) failed
A table stored in TiFlash, using this query sometimes works, sometimes not.
image
image

### PR DESCRIPTION
**Pull request by Lloyd-Pottiger**
### What problem does this PR solve?

Issue Number: ref #9472

Problem Summary:

### What is changed and how it works?

Try to fix #9472, failed to reproduce the issue. But when the filter column is the first non-empty column in the block, `block.rows() = filter_column->size()`, but filter_column is not filtered, so the rows may not match.

**But filter_column is usually added to the back of the block, so it is kind of weird.**

```commit-message
storage: fix block rows not match when the filter column is the first non-empty column in the block
```

### Check List

Tests 

- [ ] Unit test
- [ ] Integration test
- [ ] Manual test (add detailed scripts or steps below)
- [ ] No code

Side effects

- [ ] Performance regression: Consumes more CPU
- [ ] Performance regression: Consumes more Memory
- [ ] Breaking backward compatibility

Documentation

- [ ] Affects user behaviors
- [ ] Contains syntax changes
- [ ] Contains variable changes
- [ ] Contains experimental features
- [ ] Changes MySQL compatibility

### Release note

```release-note
None
```

### CODE REVIEW

File: dbms/src/Core/Block.cpp [hunk 1]
```diff
@@ -541,11 +541,9 @@ Block hstackBlocks(Blocks && blocks, const Block & header)
         return {};
 
     Block res = header.cloneEmpty();
-    size_t num_rows = blocks.front().rows();
```
Comment (JaySon-Huang) [inline dbms/src/Core/Block.cpp:544]: Suggest to add a `res.checkNumberOfRows` before returning the res

Comment (Lloyd-Pottiger) [reply dbms/src/Core/Block.cpp:544]: done

File: dbms/src/Core/Block.cpp [hunk 2]
```diff
@@ -557,6 +555,13 @@ Block hstackBlocks(Blocks && blocks, const Block & header)
     }
     res.setRSResult(rs_result);
 
+    size_t rows = res.rows();
```
[the remark this answers is not in the document]
Comment (Lloyd-Pottiger) [reply dbms/src/Core/Block.cpp]: done

File: dbms/src/Storages/DeltaMerge/LateMaterializationBlockInputStream.cpp [hunk 3]
```diff
@@ -71,8 +71,6 @@ Block LateMaterializationBlockInputStream::read()
                 }
                 for (auto & col : filter_column_block)
                 {
-                    if (col.name == filter_column_name)
```
[the remark this answers is not in the document]
Comment (Lloyd-Pottiger) [reply dbms/src/Storages/DeltaMerge/LateMaterializationBlockInputStream.cpp:74]: `col_filter` is not a ColumnRef in filter_column_block, filter_column in filter_column_block can call filter here.

File: dbms/src/Storages/DeltaMerge/LateMaterializationBlockInputStream.cpp [hunk 4]
```diff
@@ -112,12 +110,18 @@ Block LateMaterializationBlockInputStream::read()
                 // so only if the number of rows left after filtering out is large enough,
                 // we can skip some packs of the next block, call readWithFilter to get the next block.
                 rest_column_block = rest_column_stream->readWithFilter(*filter);
+                ColumnPtr filter_column;
```
[the remark this answers is not in the document]
Comment (Lloyd-Pottiger) [reply dbms/src/Storages/DeltaMerge/LateMaterializationBlockInputStream.cpp:139]: done

File: dbms/src/Storages/DeltaMerge/LateMaterializationBlockInputStream.cpp (hunk 3 above)
[the remark this answers is not in the document]
Comment (Lloyd-Pottiger) [reply dbms/src/Storages/DeltaMerge/LateMaterializationBlockInputStream.cpp:74]: Fixed

File: dbms/src/Storages/DeltaMerge/LateMaterializationBlockInputStream.cpp [hunk 5]
```diff
@@ -112,12 +133,8 @@ Block LateMaterializationBlockInputStream::read()
                 // so only if the number of rows left after filtering out is large enough,
                 // we can skip some packs of the next block, call readWithFilter to get the next block.
                 rest_column_block = rest_column_stream->readWithFilter(*filter);
-                for (auto & col : filter_column_block)
-                {
-                    if (col.name == filter_column_name)
-                        continue;
-                    col.column = col.column->filter(*filter, passed_count);
-                }
+                ColumnPtr filter_column;
```
Comment (JaySon-Huang) [inline dbms/src/Storages/DeltaMerge/LateMaterializationBlockInputStream.cpp:136]: I think we done need to create this `ColumnPtr filter_column` instance before calling `filterFilterColumnBlock`
```suggestion
```

**Review by JaySon-Huang — APPROVED** [review]
LGTM

**Pull request merged**